### PR TITLE
Fix type instability in logdet(::Matrix{Complex64})

### PR DESCRIPTION
**Pull request by jiahao**
Ref: #10188
1. Adds pi*im in a type stable fashion when there are an odd number of pivots
2. Uses mod2pi instead of mod(_, 2pi)

### CODE REVIEW

File: test/linalg/lu.jl [hunk 1]
```diff
@@ -0,0 +1 @@
+@test logdet(Complex64[1.0f0 0.5f0; 0.5f0 -1.0f0]) === 0.22314355f0 + 3.1415927f0im
```
Comment (tkelman) [inline test/linalg/lu.jl]: does this really need to be its own 1-line file?

Comment (jiahao) [inline test/linalg/lu.jl]: It is in line with @andreasnoack's work in reorganizing the linalg tests by source file. `lu.jl` simply hadn't been done yet.

Comment (tkelman) [inline test/linalg/lu.jl]: If you want to move chunks of other pre-existing tests into this file, sure that makes sense. A 1-line file is pretty pointless and not worth the noise.

Comment (StefanKarpinski) [inline test/linalg/lu.jl]: Would be worth it if there were more tests, which there should be. So maybe leave it in anticipation of more LU tests?

Comment (andreasnoack) [inline test/linalg/lu.jl]: The other LU tests are at different places in linalg1-4.jl and at some point they should go into the new lu.jl, but I don't think we should hold back this fix by requiring the reorganization to take place there. We should probably also add some more LU test.

**Comment by jiahao** [comment]
Any idea what the [build failure](https://travis-ci.org/JuliaLang/julia/jobs/50687052#L3371) means?

```
julia-debug: ast.c:716: jl_lam_gensyms: Assertion `(((jl_value_t*)((uptrint_t)((jl_value_t*)(l))->type & ((uintptr_t)~3)))==(jl_value_t*)(jl_expr_type))' failed.
```

**Comment by tkelman** [comment]
Dunno, but the fact that #10045 started happening shortly after the gensym merge sure is suspicious.

**Pull request merged**